[mlir][vector] Add unroll patterns for vector.load and vector.store

nbpatel opened the pull request:
This PR adds unroll patterns for vector.load and vector.store with rank > 1. This PR is follow up of #137558

nbpatel committed:
Add unroll patterns for vector.load and vector.store

nbpatel committed:
Merge branch 'main' into vector_unroll_load_store

nbpatel committed:
Clean up

banach-space commented on mlir/test/Dialect/Vector/vector-unroll-options.mlir:
```diff
@@ -378,3 +378,76 @@ func.func @vector_broadcast_with_tailing_unit_dim(%v: vector<4x1xf32>) -> vector
+
+// CHECK-LABEL: func.func @unroll_2D_vector_store(
+// CHECK-SAME:  %[[ARG0:.*]]: memref<4x4xf16>, %[[ARG1:.*]]: vector<4x4xf16>) {
+func.func @unroll_2D_vector_store(%arg0: memref<4x4xf16>, %arg1: vector<4x4xf16>) {
```
Please use naming consistent with other tests.
```suggestion
func.func @unroll_2D_vector_store(%mem: memref<4x4xf16>, %v: vector<4x4xf16>) {
```

banach-space reviewed (commented):
LG % some minor asks, thanks!

nbpatel committed:
Address feedback

nbpatel commented on mlir/test/Dialect/Vector/vector-unroll-options.mlir:
```diff
@@ -378,3 +378,76 @@ func.func @vector_broadcast_with_tailing_unit_dim(%v: vector<4x1xf32>) -> vector
+
+
+// CHECK-LABEL: func.func @unroll_2D_vector_load(
+// CHECK-SAME:  %[[ARG:.*]]: memref<4x4xf16>) -> vector<4x4xf16> {
```
My bad, done. Thanks

nbpatel commented on mlir/test/Dialect/Vector/vector-unroll-options.mlir:
```diff
@@ -378,3 +378,76 @@ func.func @vector_broadcast_with_tailing_unit_dim(%v: vector<4x1xf32>) -> vector
+
+// CHECK-LABEL: func.func @unroll_vector_load(
+// CHECK-SAME:  %[[ARG:.*]]: memref<4x4x4x4xf16>) -> vector<2x2xf16> {
+func.func @unroll_vector_load(%arg0: memref<4x4x4x4xf16>) -> vector<2x2xf16> {
```
This is another variant where we load a 2D block from 4D memref...added this just to test the offset computation. changed the name of the test.

nbpatel commented on mlir/lib/Dialect/Vector/Transforms/VectorUnroll.cpp:
```diff
@@ -54,6 +54,33 @@ static SmallVector<Value> sliceTransferIndices(ArrayRef<int64_t> elementOffsets,
+// compute the new indices for vector.load/store by adding offsets to
+// originalIndices.
+// It assumes m <= n (m = offsets.size(), n = originalIndices.size())
+// Last m of originalIndices will be updated.
```
done

nbpatel commented:
Thanks @banach-space for the feedback. I addressed your feedback. Please take a look again whenever you have some time.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorUnroll.cpp:
```diff
@@ -54,6 +54,33 @@ static SmallVector<Value> sliceTransferIndices(ArrayRef<int64_t> elementOffsets,
 
+// Compute the new indices for vector.load/store by adding `offsets` to
+// `originalIndices`.
+// It assumes m <= n (m = offsets.size(), n = originalIndices.size())
+// Last m of `originalIndices` will be updated.
```
If the purpose of the method is to add two containers, then 
```suggestion
// If m < n (m = offsets.size(), n = originalIndices.size()), then only the trailing m values in `originalIndices` are updated.
```

This will be sufficient to communicate that `n >= m`. And then, `assert` inside the method can verify that.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorUnroll.cpp:
```diff
@@ -631,6 +658,123 @@ struct UnrollGatherPattern : public OpRewritePattern<vector::GatherOp> {
   vector::UnrollVectorOptions options;
 };
 
+// clang-format off
```
Let's avoid `clang-format off` and instead try to trim the comment to 80 cols. You have multiple options:
```cpp
// OPTION 1 (split across mulitple lines)
// %result_0 = vector.insert_strided_slice %slice_0, %cst 
//     {offsets = [0, 0], strides = [1]} 
//     : vector<4xf32> into vector<4x4xf32>
//
// OPTION 2 (skip some info, shorten names)
// %s0 = vector.load %base[%indices] : memref<4x4xf32>, vector<4xf32>
// %res_0 = vector.insert_strided_slice %s0, %cst  {offsets = [0, 0], strides = [1]} ...
```

banach-space reviewed (commented):
Hey, thanks for the update!

On second scan, it feels that `computeIndices` could be simplified. Please take a look. Thank you!

nbpatel committed:
Simplify computeIndices

banach-space reviewed (approved):
LGTM, thanks for addressing all my comments, the final version is much easier to follow 🙏🏻 

@newling , IIRC these patterns were of interest to you as well? @nbpatel , would you mind waiting ~1 day before landing (to give others a chance to take a look)?

newling commented on mlir/test/lib/Dialect/Vector/TestVectorTransforms.cpp:
```diff
@@ -178,6 +178,16 @@ struct TestVectorUnrollingPatterns
 
+    populateVectorUnrollPatterns(
+        patterns, UnrollVectorOptions()
+                      .setNativeShape(ArrayRef<int64_t>{2, 2})
```
I think {2,2} should be used

nbpatel commented:
> Thanks for moving this from linearization to unrolling, i.e. from #136193 and #137558
> 
> As far as I can tell the existing 9 unroll patterns support unrolling to any tile size of the vector. See for example
> 
> elementwise:
> 
> https://github.com/llvm/llvm-project/blob/13fe07d670e8a115929c9e595c4490ef5c75f583/mlir/test/Dialect/Vector/vector-transforms.mlir#L187
> 
> transfer_read:
> 
> https://github.com/llvm/llvm-project/blob/13fe07d670e8a115929c9e595c4490ef5c75f583/mlir/test/Dialect/Vector/vector-transfer-unroll.mlir#L29
> 
> and the tests in the file you added your tests to in this PR. Can you motivate deviating from this design? For example, I would expect based on the other unroll patterns for provide a 'target shape' of 2x2 to convert a load of 6x6 to 9 loads of 2x2.
> 
> I think that transfer_read is a more complex op than vector load. For example, transfer_read is lowered to load [here](https://github.com/llvm/llvm-project/blob/f12b1ed11672bc40a53fb1180541b2fda6e7d9fc/mlir/lib/Dialect/Vector/Transforms/LowerVectorTransfer.cpp#L406). So I think it should be possible to unroll load to any tile size, if that's what happens for transfer_reads.
> 
> In terms of achieving what is I think your goad of only having rank-1 loads, I propose the following
> 
> 1. use unrolling to go from loads of `6x6` to 6 loads of `1x6`
> 2. linearize the loads of `1x6` to loads of `6`
> 
> Inn my opinion, having a linearization pattern for 2 above is fine, because it converts **one** rank-2 op into **one** rank-1 op. That's what I was kind of driving at when I added this comment:
> 
> https://github.com/llvm/llvm-project/blob/a53003fe23cb6c871e72d70ff2d3a075a7490da2/mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h#L411
> 
> I acknowledge that end result is exactly what you had in #136193 !! But I think it makes Vector more modular and intuitive for other users if unrolling and linearization have a clear boundary.

The motivation for this unrolling was to have 1D loads and stores and hence the target shape is set accordingly. But, I agree with your point of making it more modular to align with vector dialect's flexible unrolling design. Can I address this in a follow up PR?

nbpatel committed:
Use unroll options

nbpatel committed:
Formatting

newling reviewed (commented):
Thanks @nbpatel, this looks good to me! My remaining comments are all quite minor

nbpatel committed:
Address comments

nbpatel commented:
> Thanks @nbpatel, this looks good to me! My remaining comments are all quite minor

Thanks, I addressed your comments. Can you please approve if it looks ok?

newling reviewed (approved):
LGTM
